### ISSUE TITLE
cross entropy loss

### ISSUE THREAD

**Issue by trungpham2606** [rpautrat/SuperPoint#114]
@rpautrat Thank you for your great implementation !
Iam doing this model on Pytorch. But I dont get the idea about the cross-entropy loss in this SuperPoint paper. Can you give me some info about the shape of groundtruth point ? is its shape (65, h, w) and 65 stands for [1, 1, 1, ...., 0] (for example) ? 
I tried to implement the loss according to the paper but it didnt success.

**Comment by trungpham2606** [comment]
@rpautrat tks for your prompt response. 
I have a question about the normalization factor in descriptor loss. Why did you multiply the valid_mask with Hc and Wc. While training, my normalization were high all the time so the loss is super small. 
Can you give me the idea behind that equation ?

**Comment by rpautrat** [comment]
In the original paper, the loss is divided by Hc²xWc², because there are two double sums, each of them in [1; Hc] and [1; Wc].

In this implementation, the first double sum (on the centers in the first image) is still there, but the second sum is actually smaller. We indeed mask the pixels of the second (warped image) that are not present in the first image. So instead of normalizing this second sum by HcxWc, we normalize it by `tf.reduce_sum(valid_mask)`, in other words the number of points that are actually not masked out.

Hence, the final normalization looks like this:
```
normalization = nb_elements_in_first_sum * nb_elements_in_second_sum = tf.to_float(Hc * Wc) * tf.reduce_sum(valid_mask)
```

**Comment by trungpham2606** [comment]
@rpautrat  Can you show me the descriptor loss during training. Without loading the pretrained weights, the losses are still so small. exp: 0.006 ?

**Comment by trungpham2606** [comment]
@rpautrat 
![image](https://user-images.githubusercontent.com/35657426/67559297-e0077c80-f742-11e9-90d1-123f3639f067.png)
and in your descriptor_loss function.
the shape of valid_mask after using tf.reduce_prod should be (batch_size, Hc, Wc, 1) right ?

**Comment by rpautrat** [comment]
Yes sorry, the shape after tf.reduce_prod should actually be (batch_size, Hc, Wc).

Here is the my descriptor loss:
![descriptor_loss](https://user-images.githubusercontent.com/32239569/67566866-0b7d6d80-f729-11e9-9c74-69c0d50f986c.png)

So more at about 0.08 actually. This is with the default parameter `lambda_loss = 1`. 0.006 seems a bit low, maybe you can try to increase your `lamda_loss` parameter.

**Comment by trungpham2606** [comment]
@rpautrat  Iam implementing by pytorch according to your tf version. Now Iam freezing all the layers except the layers related to descriptor head then training. The loss function now is descriptor loss only. But I still didnt figure out why whether I used the pre-trained weights or not, the loss is always super low.

**Comment by rpautrat** [comment]
I see, then if you froze the backbone layers, it makes sense that your loss is smaller (the backbone was not frozen in my case).

There is absolutely no difference in the loss at the beginning of training when you compare with/without pre-trained weights?

**Comment by trungpham2606** [comment]
@rpautrat  yeah. The losses in 2 cases are nearly the same. In addition, i didnt multiply Hc * Wc in normalization as in your equation. If I added them, the loss will be 1200 times smaller then. (0.006/1200)
That's why I asked you about the term (Hc*Wc) in normalization but then I realized that your equation is followed by the author's one.

**Comment by rpautrat** [comment]
I see. But this small loss might be normal when training with the descriptors only. Have you tried to evaluate its performance?

**Comment by trungpham2606** [comment]
@rpautrat No not yet cause iam still so confused about the loss value. 1 more thing I realized that in pytorch pretrained weights from Magic Leap, it doesnt have batch normalization. ? But in their paper, they stated that they used batch normalization and relu after each convolutional layer.

**Comment by rpautrat** [comment]
I think I also tried with and without batch normalization, but there is no significant difference anyway.

**Comment by ddetone** [comment]
The Magic Leap model was trained with batch norm. We folded the batch norm layers out after training by combining them with the convolution layers in the released model.

**Comment by trungpham2606** [comment]
@ddetone so according to your response that I dont have to add bn layer after each conv layer, right ? Just loading the pretrained weights then train the model normally ? 
@rpautrat Iam following your tf descriptor loss but the results are so bad :-S

**Comment by trungpham2606** [comment]
@rpautrat i saw on your code, in descriptor loss calculation part, the desc features were not normalized yet, right ?

**Comment by rpautrat** [comment]
Not on master, no. I am still investigating what is the problem in the descriptors loss. Even after normalizing, the loss has not the expected behavior, even on simple handcrafted examples (see https://github.com/rpautrat/SuperPoint/issues/95#issuecomment-540535549).

**Comment by trungpham2606** [comment]
@rpautrat When I tried 2 versions of function warp points, one of yours and the other is mine. The results are quite different. 
Here is the results from my function (just using cv2.perspective(pts, homo_matrix)
![image](https://user-images.githubusercontent.com/35657426/68012384-747e5b80-fcbc-11e9-9fac-1ce6c18947a3.png)
The points on left and right hand side of the image are correct.
But when I run your code, the results are different a bit.
![image](https://user-images.githubusercontent.com/35657426/68012467-b4ddd980-fcbc-11e9-9f7b-1cb3cb5f3508.png)

Iam quite confused about these results.

**Comment by rpautrat** [comment]
This might explain the bug: https://github.com/rpautrat/SuperPoint/issues/117#issue-515271640

**Comment by trungpham2606** [comment]
@rpautrat Iam using the Hpatches dataset for training, but I can observe that there are some wrong homography matrix in this dataset as well as some images dont have the same size (in same folder). Iam training this network with new strategies and the results are way better than the first version which I follow your tensorflow code.

**Comment by rpautrat** [comment]
Yes, I noticed it too. I have fixed the wrong homographies and I am running again the evaluations, but I will push the corrected version today on branch superpoint_v1.

**Comment by trungpham2606** [comment]
@rpautrat and when I test your pretrained weights, I noticed that your model cant predict well when the image was rotated more than 45 degrees. Both your work and Magicleap (magic leap's results seem to be worse, it doesnt show any matching pairs). But after training, I can generate so many matching pair whatever the rotation angles are.
My result:
![image](https://user-images.githubusercontent.com/35657426/68372188-f239df80-0172-11ea-8173-201f17850bdc.png)
Magic Leap's result:
![image](https://user-images.githubusercontent.com/35657426/68372250-12699e80-0173-11ea-9012-18e68a8d8030.png)
Your result is the same as MagicLeap's
I have a question that either you didnt train on more than 45 degrees or after training the model still cant work well on rotation angles which are greater than 45 ?

**Comment by rpautrat** [comment]
During training, I used angles of maximum 90 degrees, because the network was struggling to learn with big rotations at the time. But this was some time ago, many improvements have been made since, so I could try again to train with 180 degrees rotations.

**Comment by trungpham2606** [comment]
@rpautrat Iam looking forward to your result. Iam still struggling with training because I have 1 PC only for training this network :((( but with this very first results, I think this network can work well with any rotation angles.

**Comment by rpautrat** [comment]
> @rpautrat Iam using the Hpatches dataset for training, but I can observe that there are some wrong homography matrix in this dataset as well as some images dont have the same size (in same folder). Iam training this network with new strategies and the results are way better than the first version which I follow your tensorflow code.

The fix is now pushed to superpoint_v1, with the corresponding new evaluation.

**Comment by trungpham2606** [comment]
@rpautrat lets try retrain with rotation angles greater than 180 degree. Iam looking forward to your results.

**Comment by Vincentqyw** [comment]
> The Magic Leap model was trained with batch norm. We folded the batch norm layers out after training by combining them with the convolution layers in the released model.

Hi @rpautrat  , I noticed that your implementation is **conv+relu+bn**,  which is cannot be merge bn layers and conv layers into conv layers. Compared with the release model of @ddetone , I think he put bn layer before relu, only in this way, he can merge the two layers together.

**Comment by rpautrat** [comment]
Technically the original paper states the following:

> All convolution layers in the network are followed by ReLU non-linear activation and BatchNorm normalization.

which I interpret as **conv+relu+bn**. But in the released model, it could be that it was the other configuration that was used.
